CUDADataFormats/SiPixelCluster/interface/SiPixelClustersCUDA.h

```diff
@@ -0,0 +1,57 @@
+#ifndef CUDADataFormats_SiPixelCluster_interface_SiPixelClustersCUDA_h
+#define CUDADataFormats_SiPixelCluster_interface_SiPixelClustersCUDA_h
+
+#include "CUDADataFormats/Common/interface/device_unique_ptr.h"
+
+#include <cuda/api_wrappers.h>
+
+class SiPixelClustersCUDA {
+public:
+  SiPixelClustersCUDA() = default;
+  explicit SiPixelClustersCUDA(size_t feds, size_t nelements, cuda::stream_t<>& stream);
+  ~SiPixelClustersCUDA() = default;
+
+  SiPixelClustersCUDA(const SiPixelClustersCUDA&) = delete;
+  SiPixelClustersCUDA& operator=(const SiPixelClustersCUDA&) = delete;
+  SiPixelClustersCUDA(SiPixelClustersCUDA&&) = default;
+  SiPixelClustersCUDA& operator=(SiPixelClustersCUDA&&) = default;
+
+  uint32_t * __restrict__ moduleStart() { return moduleStart_d.get(); }
+  int32_t  * __restrict__ clus() { return clus_d.get(); }
+  uint32_t * __restrict__ clusInModule() { return clusInModule_d.get(); }
+  uint32_t * __restrict__ moduleId() { return moduleId_d.get(); }
+  uint32_t * __restrict__ clusModuleStart() { return clusModuleStart_d.get(); }
+
+  uint32_t const * __restrict__ moduleStart() const { return moduleStart_d.get(); }
+  int32_t  const * __restrict__ clus() const { return clus_d.get(); }
+  uint32_t const * __restrict__ clusInModule() const { return clusInModule_d.get(); }
+  uint32_t const * __restrict__ moduleId() const { return moduleId_d.get(); }
+  uint32_t const * __restrict__ clusModuleStart() const { return clusModuleStart_d.get(); }
+
+  uint32_t const * __restrict__ c_moduleStart() const { return moduleStart_d.get(); }
+  int32_t  const * __restrict__ c_clus() const { return clus_d.get(); }
+  uint32_t const * __restrict__ c_clusInModule() const { return clusInModule_d.get(); }
+  uint32_t const * __restrict__ c_moduleId() const { return moduleId_d.get(); }
+  uint32_t const * __restrict__ c_clusModuleStart() const { return clusModuleStart_d.get(); }
+
+  struct DeviceConstView {
+    uint32_t const *moduleStart;
+    int32_t  const *clus;
+    uint32_t const *clusInModule;
+    uint32_t const *moduleId;
+    uint32_t const *clusModuleStart;
+  };
```

Comment:
I replaced the "me" pointer for digis+clusters with a "view" that gets passed to the kernels by value (currently used only in `SimTracker/TrackerHitAssociation/plugins/ClusterSLOnGPU.cu`). In principle we could use this pattern also in place of pointers, but in the end I decided to not try that in this PR.

If we want to have the "me" pointer instead, I believe the options are the following
* forget per-event device allocations
* add caching allocator for pinned host memory as well
   * just to remind, the pinned memory is needed for asynchronous transfers
* go to unified memory

Reply:
The downside is that every member pointer eats up the kernel launch argument stack. According to random results from google one can pass at most 256 bytes to a kernel, corresponding to 32 pointers.

In addition, would the struct need accessors with the  `__restricted__` decoration? (I didn't try to do it here because the "me" pointed-to-struct didn't have them either)

Reply:
the __restricted__ decoration is apparently useless (see my other thread)
it works inly in the signature of the __global__ function for pointers passed directly to the kernel.
in all other cases one has to use the _ldg intrinsics.
```
T x(int i) const { return __ldg(x_d+i);}
```
etc

Reply:
Thanks. So the effect here would be just dropping the `__restrict__` as the function calls are to be done on the host side, right?

(and leave the `__ldg` variants to the "device view" if the start to apply that pattern, or with the "me" pointers)